### ISSUE TITLE
Suspecious bug

### ISSUE THREAD

**Issue by Jackberg** [facebookresearch/ParlAI#609]
I've noticed an abnormal phenomenon for seq2seq model: the validation loss is always lower than training loss, both on Twitter and OpenSubtitles dataset.

And I found this part of codes quite suspicious:
https://github.com/facebookresearch/ParlAI/blob/09351eec1076319a9f5ea0f6a0c242d5f35fa43d/parlai/agents/seq2seq/seq2seq.py#L403-L411

So here we see when doing pure prediction, we don't use true labels `ys`, but when computing loss we use, which I think will bias the validation loss. Are there any reasons for computing loss like this?

**Comment by alexholdenmiller** [comment]
Well, during training it has dropout enabled which should hurt its loss.

During validation, you're calculating the average loss on each token, so you do want to feed in the correct token not the one generated at the previous time step.

So both of these are expected!

**Comment by Jackberg** [comment]
I think you're right @alexholdenmiller . Do you know whether the CrossEntropyLoss is averaged over sequence length or not? I'm not sure about this from the PyTorch documentation.

**Comment by alexholdenmiller** [comment]
By default it is, but `size_average=False` disables that. A better implementation here would set size average to false and then sum the loss over all tokens in the entire validation set, then average, then calculate perplexity. I'll keep this issue open as a reminder to implement that instead of calculating perplexity per minibatch and averaging those.

**Comment by Jackberg** [comment]
Well, I'm not sure about the `loss1` by averaging `total loss` over `total # words`, but I would say `loss2` (averaged by `total # words per batch`) is an unbiased estimation of the `loss1`. So I think it is not a bug of this implementation?

**Comment by Jackberg** [comment]
And thanks @alexholdenmiller ! With `dropout=0.0`, the validation loss is roughly the same as training loss.

**Comment by alexholdenmiller** [comment]
Great!

re the averaging: it doesn't matter for the loss, but for the perplexity the exponent gives you a different result if you average per minibatch or over the whole set

**Comment by alexholdenmiller** [comment]
Updated perplexity scoring
